src/rdsn/src/meta/server_state.h

```diff
@@ -353,6 +353,18 @@ class server_state
                                                   const std::vector<std::string> &keys,
                                                   const std::vector<std::string> &values);
 
+    bool app_info_compatible_equal(const app_info &l, const app_info &r) const
+    {
+        if (l.status != r.status || l.app_type != r.app_type || l.app_name != r.app_name ||
```

Comment:
Better to add some comments to explain why do this partial fields check.

Reply:
> // compatible for app.duplicating different between primary and secondaries in 2.1.x, 2.2.x and 2.3.x release

The comment is added for its caller, it will only be used for this special compatible case.